### DISCUSSION TITLE
How to pick the best preconditioner and how to analyze the properties of mfem::Operator?

### DISCUSSION THREAD

**Discussion by kovar** [mfem/mfem#3358] [Q&A] [answered]
Hello 👋 

I've recently started working on a project from academia that uses MFEM and my task is to analyze the preconditioners being used and to figure out which one is the fastest for the given linear problem that is being solved (Navier-Stokes equations on 2D incompressible flow around hydrofoils). After reading some literature it seems that in order to make an informed decision on a preconditioner, it would be useful to know the properties of $A$, for example sparsity, positive definiteness or diagonal dominance. I understand that this $A$ is essentially `mfem::Operator`, right? I see there is a method called [PrintMatlab](https://docs.mfem.org/html/classmfem_1_1Operator.html#a4ade1793b4efc7060337795f07ab28c9), but after trying that it gives only `NaN`'s. But this seems infeasible anyway as the matrix is too large to store in memory.

The project also implements a `Mult()` method, so I tried "poking" the operator with simple unit vectors, but that is not working so far either. I don't think I am allowed to share this code here unfortunately, so I am looking for alternatives.

Is there any way that one could see the properties of `mfem::Operator` (i.e. $A$ in a $A\mathbf{x} = \mathbf{b}$) from MFEM? Or how else should I go about picking the best preconditioner for a given linear problem other than trial & error?

Thank you in advance for any help! I am very much a beginner, so any advice will help!

**Comment by pazner** [comment, marked as answer]
Hello @kovar,

Usually the best way to choose a preconditioner is based on theory or literature. Since the incompressible Navier-Stokes equations are very well-studied, there are many papers discussing appropriate choices of preconditioners. It will depend strongly on exactly the formulation you are using to discretize/solve your problem. If you are using a projection/split method, the systems you will need to solve will be diffusion-like, and multigrid preconditioners (e.g. AMG) will work well.

**Reply by kovar** [reply]
Thank you, good to know!
